[JsonPath] Add documentation (#21078)
symfony/symfony-docs: components/json_path.rst

The JsonPath Component
======================

.. versionadded:: 7.3

    The JsonPath component was introduced in Symfony 7.3 as an
    :doc:`experimental feature </contributing/code/experimental>`.

The JsonPath component provides a powerful way to query and extract data from

> JSON structures. It implements the `RFC 9535 (JSONPath) <https://datatracker.ietf.org/doc/html/rfc9535>`_

Reviewer: Link are added at end of file.
See `http_cache.rst` for `RFC 7234` for example

```suggestion
JSON structures. It implements the `RFC 9535 - JSONPath`_
```

And add this line at end of file
```
.. _`RFC 9535 - JSONPath`: https://tools.ietf.org/html/rfc7234
```

Author: Done with the good link

standard, allowing you to navigate complex JSON data with ease.

Just as the :doc:`DomCrawler component </components/dom_crawler>` allows you to navigate and query HTML/XML documents
using XPath, the JsonPath component provides a similar experience to traverse and search JSON structures
using JSONPath expressions. The component also offers an abstraction layer for data extraction.

Installation
------------

You can install the component in your project using Composer:

.. code-block:: terminal

    $ composer require symfony/json-path

.. include:: /components/require_autoload.rst.inc

Usage
-----

To start querying a JSON document, first create a :class:`Symfony\\Component\\JsonPath\\JsonCrawler`
object from a JSON string. For the following examples, we'll use this sample
"bookstore" JSON data::

>     use Symfony\Component\JsonPath\JsonCrawler;

Reviewer: I created https://github.com/symfony-tools/symfony-application/pull/36 in order to fix code-block-checker false positive

> [Missing class] Class, interface or trait with name "Symfony\Component\JsonPath\JsonCrawler" does not exist

Reviewer 2: Merged

Author: Thanks @alamirault and @OskarStark!

Searching with the Descendant Operator
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The descendant operator (``..``) recursively searches for a given key, allowing
you to find values without specifying the full path.

::

    // Get all authors from anywhere in the document
    $authors = $crawler->find('$..author');

    // $authors is ['Nigel Rees', 'Evelyn Waugh', 'Herman Melville', 'John Ronald Reuel Tolkien']

Filtering Results
~~~~~~~~~~~~~~~~~

JSONPath includes a powerful filter syntax (``?(<expression>)``) to select items
based on a condition. The current item within the filter is referenced by ``@``.

::

    // Get all books with a price less than 10
    $cheapBooks = $crawler->find('$.store.book[?(@.price < 10)]');

    /*
    $cheapBooks contains two book objects:
    the one by "Nigel Rees" and the one by "Herman Melville"
    */

Building Queries Programmatically
---------------------------------

For more dynamic or complex query building, you can use the fluent API provided
by the :class:`Symfony\\Component\\JsonPath\\JsonPath` class. This allows you
to construct a query object step-by-step. The ``JsonPath`` object can then be

> passed to the crawler's method:`Symfony\\Component\\JsonPath\\JsonCrawler::find` method.

Reviewer: ```suggestion
passed to the crawler's :method:`Symfony\\Component\\JsonPath\\JsonCrawler::find` method.
```

Author: Done

The main advantage of the programmatic builder is that it automatically handles
the correct escaping of keys and values, preventing syntax errors.

::

    use Symfony\Component\JsonPath\JsonPath;

    $path = (new JsonPath())
        ->key('store')       // Selects the 'store' key
        ->key('book')        // Then the 'book' key
        ->index(1);          // Then the item at index 1 (the second book)

>     // The created $path object is equivalent to the string '$["store"]["book"][1]'

Reviewer: Looks good, however the only syntax we explained above is `$.store.book[1]`. We should maybe show this alternative first ? Maybe ~L94

Author: I suggested something, is that all right with you?

The JsonPath Component
======================

.. versionadded:: 7.3

    The JsonPath component was introduced in Symfony 7.3 as an
    :doc:`experimental feature </contributing/code/experimental>`.

The JsonPath component provides a powerful way to query and extract data from
JSON structures. It implements the `RFC 9535 (JSONPath) <https://datatracker.ietf.org/doc/html/rfc9535>`_
standard, allowing you to navigate complex JSON data with ease.

Just as the :doc:`DomCrawler component </components/dom_crawler>` allows you to navigate and query HTML/XML documents

> using CSS selectors, the JsonPath component provides a similar fluent interface to traverse and search JSON structures

Reviewer 3: I think we should mention XPath instead of CSS selector. JsonPath is directly inspired from XPath so it makes more sense

> using CSS selectors, the JsonPath component provides a similar fluent interface to traverse and search JSON structures

Reviewer 3: JsonPath doesn't provide fluent interfaces, so I think this should reworded as well

> using JSONPath expressions. It's component also offer an abstraction layer for data extraction.

Reviewer 3: ```suggestion
using JSONPath expressions. The component also offers an abstraction layer for data extraction.
```

Installation
------------

You can install the component in your project using Composer:

.. code-block:: terminal

    $ composer require symfony/json-path

.. include:: /components/require_autoload.rst.inc

Usage
-----

To start querying a JSON document, first create a :class:`Symfony\\Component\\JsonPath\\JsonCrawler`
object from a JSON string. For the following examples, we'll use this sample
"bookstore" JSON data:

> ::

Reviewer 3: ```suggestion
"bookstore" JSON data::
```

    $json = <<<'JSON'
    {
        "store": {
            "book": [
                {
                    "category": "reference",
                    "author": "Nigel Rees",
                    "title": "Sayings of the Century",
                    "price": 8.95
                },
                {
                    "category": "fiction",
                    "author": "Evelyn Waugh",
                    "title": "Sword of Honour",
                    "price": 12.99
                },
                {
                    "category": "fiction",
                    "author": "Herman Melville",
                    "title": "Moby Dick",
                    "isbn": "0-553-21311-3",
                    "price": 8.99
                },
                {
                    "category": "fiction",
                    "author": "John Ronald Reuel Tolkien",
                    "title": "The Lord of the Rings",
                    "isbn": "0-395-19395-8",
                    "price": 22.99
                }
            ],
            "bicycle": {
                "color": "red",
                "price": 399
            }
        }
    }
    JSON;

    $crawler = new JsonCrawler($json);

> Once you have the crawler instance, use its ``find()`` method to start querying

Reviewer 3: You can use the :method: directive

the data. This method always returns an array of matching values.

Querying with Expressions
-------------------------

The primary way to query the JSON is by passing a JSONPath expression string
to the ``find()`` method.

> **Accessing a Specific Property**

Reviewer 3: Subtitles use tildes `~` as you'll see in other pages

Use dot-notation for object keys and square brackets for array indices. The root
of the document is represented by ``$``.

> ::

Reviewer 3: ```suggestion
of the document is represented by ``$``::
```

                    "price": 22.99
                }
            ],
            "bicycle": {
                "color": "red",
                "price": 399
            }
        }
    }
    JSON;

    $crawler = new JsonCrawler($json);

Once you have the crawler instance, use its :method:`Symfony\\Component\\JsonPath\\JsonCrawler::find` method to start querying
the data. This method always returns an array of matching values.

Querying with Expressions
-------------------------

The primary way to query the JSON is by passing a JSONPath expression string
to the :method:`Symfony\\Component\\JsonPath\\JsonCrawler::find` method.

Accessing a Specific Property
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Use dot-notation for object keys and square brackets for array indices. The root
of the document is represented by ``$``::

    // Get the title of the first book in the store
    $titles = $crawler->find('$.store.book[0].title');

    // $titles is ['Sayings of the Century']

While dot-notation is common, JSONPath offers alternative syntaxes which can be more flexible. For instance, bracket notation (`['...']`) is required if a key contains spaces or special characters.

> ::

Reviewer 3: ```suggestion
While dot-notation is common, JSONPath offers alternative syntaxes which can be more flexible. For instance, bracket notation (`['...']`) is required if a key contains spaces or special characters::
```

Same at other places of the doc

Author: Fixed

    /*
    $cheapBooks contains two book objects:
    the one by "Nigel Rees" and the one by "Herman Melville"
    */

Building Queries Programmatically
---------------------------------

For more dynamic or complex query building, you can use the fluent API provided
by the :class:`Symfony\\Component\\JsonPath\\JsonPath` class. This allows you
to construct a query object step-by-step. The ``JsonPath`` object can then be
passed to the crawler's :method:`Symfony\\Component\\JsonPath\\JsonCrawler::find` method.

The main advantage of the programmatic builder is that it automatically handles
the correct escaping of keys and values, preventing syntax errors.

::

    use Symfony\Component\JsonPath\JsonPath;

    $path = (new JsonPath())
        ->key('store')       // Selects the 'store' key
        ->key('book')        // Then the 'book' key
        ->index(1);          // Then the item at index 1 (the second book)

    // The created $path object is equivalent to the string '$["store"]["book"][1]'
    $book = $crawler->find($path);

    // $book contains the book object for "Sword of Honour"

The ``JsonPath`` class provides several methods to build your query:

> ``key(string $name)``

Reviewer 3: If I remember correctly, you should use the list notation using `*`. You can find many places where it is done this way in the other pages of the doc

Author: Fixed
